### PR TITLE
Lon 959

### PR DESCRIPTION
**Pull request by vmartos**

### CODE REVIEW

File: LongoMatch.Services/Controller/TeamsController.cs [hunk 1]
```diff
@@ -57,6 +58,18 @@ protected override bool SaveValidations (Team model)
 		{
 			return true;
 		}
+
+		protected override void HandleTemplateChanged (object sender, System.ComponentModel.PropertyChangedEventArgs e)
```
Comment (vguzmanp) [inline LongoMatch.Services/Controller/TeamsController.cs]: using

File: LongoMatch.Services/Controller/TeamsController.cs [hunk 2]
```diff
@@ -57,6 +58,18 @@ protected override bool SaveValidations (Team model)
...
+		{
+			base.HandleTemplateChanged (sender, e);
+			if (e.PropertyName == "Model") {
+				var viewModel = ViewModel as TeamsManagerVM;
+				if (viewModel != null) {
```
Comment (vguzmanp) [inline LongoMatch.Services/Controller/TeamsController.cs]: use `NeedsSync`

File: LongoMatch.Services/ViewModel/NewProjectPanelVM.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
+﻿//
+//  Copyright (C) 2017 Fluendo S.A.
+using System;
+using LongoMatch.Core.Store;
+using LongoMatch.Core.ViewModel;
+using LongoMatch.Services.Interfaces;
+
+namespace LongoMatch.Services.ViewModel
+{
+	public class NewProjectPanelVM : LMProjectVM, ILMTeamTaggerVM
```
Comment (vguzmanp) [inline LongoMatch.Services/ViewModel/NewProjectPanelVM.cs]: I think this should be called `NewProjectVM`, shouldn't it? `Panel` is a kind of View...

File: LongoMatch.Services/State/NewProjectState.cs [hunk 4]
```diff
@@ -35,10 +36,10 @@ public class NewProjectState : ScreenState<IViewModel>
 
 		protected override void CreateViewModel (dynamic data)
 		{
-			if (data == null) {
-				data = new LMProjectVM ();
+			ViewModel = new NewProjectPanelVM ();
+			if (data != null) {
+				ViewModel.Model = data;
```
Comment (vguzmanp) [inline LongoMatch.Services/State/NewProjectState.cs]: I prefer to send VM in the `data` instead of Models. Why the change?

File: LongoMatch.Services/ViewModel/LMProjectAnalysisVM.cs [hunk 5]
```diff
@@ -48,5 +82,19 @@ public LMProjectAnalysisVM ()
...
 		}
+
+		void ResetTeamTagger ()
+		{
+			TeamTagger.AwayTeam = Project.AwayTeam;
+			TeamTagger.HomeTeam = Project.HomeTeam;
+			TeamTagger.Background = Project.Dashboard.Model != null ? Project.Dashboard.FieldBackground : null;
+		}
+
+		void HandleProjectPropertyChanged (object sender, System.ComponentModel.PropertyChangedEventArgs e)
```
Comment (vguzmanp) [inline LongoMatch.Services/ViewModel/LMProjectAnalysisVM.cs]: using

File: LongoMatch.Services/ViewModel/LMTeamEditorVM.cs [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+using LongoMatch.Core.Store;
+using LongoMatch.Core.ViewModel;
+using VAS.Core;
+using VAS.Core.Common;
+using VAS.Core.Events;
+using VAS.Core.MVVMC;
+
+namespace LongoMatch.Services.ViewModel
+{
+	public class LMTeamEditorVM : ViewModelBase
```
Comment (vguzmanp) [inline LongoMatch.Services/ViewModel/LMTeamEditorVM.cs]: Add comment for the class

File: LongoMatch.Services/ViewModel/PlayEditorVM.cs [hunk 7]
```diff
@@ -15,28 +15,104 @@
...
+		/// </summary>
+		/// <value>The team tagger.</value>
+		public LMTeamTaggerVM TeamTagger {
+			get;
+			protected set;
+		}
+
+		public LMProjectVM Project {
+			get {
+				return this;
```
Comment (vguzmanp) [inline LongoMatch.Services/ViewModel/PlayEditorVM.cs]: Why do you need this?

File: LongoMatch.Services/ViewModel/PlayEditorVM.cs [hunk 8]
```diff
@@ -15,28 +15,104 @@
...
+		{
+			base.SyncLoadedModel ();
+			ResetTeamTagger ();
+		}
+
+		void ResetTeamTagger ()
+		{
+			TeamTagger.AwayTeam = AwayTeam;
+			TeamTagger.HomeTeam = HomeTeam;
+			TeamTagger.Background = Model.Dashboard?.FieldBackground;
```
Comment (vguzmanp) [inline LongoMatch.Services/ViewModel/PlayEditorVM.cs]: Having this Background as null will crash somewhere else, won't it?

File: LongoMatch.Services/ViewModel/PlayEditorVM.cs [hunk 9]
```diff
@@ -15,28 +15,104 @@
...
+			UpdateTeamTagger ();
+		}
+
+		void UpdateTeamTagger ()
+		{
+			if (play != null) {
```
Comment (vguzmanp) [inline LongoMatch.Services/ViewModel/PlayEditorVM.cs]: Use a [guard clause](https://fluendo.atlassian.net/wiki/display/Team/Coding+Style#CodingStyle-Guardclauses). Also, don't reference the internal variable, reference the property

File: LongoMatch.Services/ViewModel/PlayEditorVM.cs [hunk 10]
```diff
@@ -15,28 +15,104 @@
...
+				foreach (var player in TeamTagger.HomeTeam) {
+					if (play.Players.Contains (player.Model)) {
```
Comment (vguzmanp) [inline LongoMatch.Services/ViewModel/PlayEditorVM.cs]: Why can't you use the ViewModels?

File: LongoMatch.Services/ViewModel/SubstitutionsEditorVM.cs [hunk 11]
```diff
@@ -15,17 +15,132 @@
...
+			} else {
+				await App.Current.EventsBroker.Publish (new UpdateEvent<SubstitutionEvent> {
+					Object = play as SubstitutionEvent
+				});
+			}
+		}
+
+		void UpdateViewModels ()
+		{
+			if (play != null && Model != null) {
```
Comment (vguzmanp) [inline LongoMatch.Services/ViewModel/SubstitutionsEditorVM.cs]: use `Play` instead of `play`

File: Tests/Controller/TestLMTeamTaggerController.cs [hunk 12]
```diff
@@ -0,0 +1,330 @@
...
+		{
+			ProjectSetUp ();
+
+			Assert.AreEqual (4, teamTagger.HomeTeam.FieldPlayersList.Count ());
+			Assert.AreEqual (1, teamTagger.HomeTeam.BenchPlayersList.Count ());
+
+			teamTagger.HomeTeam.Model.FormationStr = "2-1";
+
+			Assert.AreEqual (3, teamTagger.HomeTeam.FieldPlayersList.Count ());
+			Assert.AreEqual (2, teamTagger.HomeTeam.BenchPlayersList.Count ());
```
Comment (vguzmanp) [inline Tests/Controller/TestLMTeamTaggerController.cs]: What's the difference between this and the previous one? How is it different with project?

File: Tests/Services/ViewModel/TestTeamsManagerVM.cs [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+using System;
+using LongoMatch.Core.ViewModel;
+using LongoMatch.Services.ViewModel;
+using NUnit.Framework;
+using VAS.Core.Common;
+
+namespace Tests.Services.ViewModel
+{
+	[TestFixture]
+	public class TestTeamsManagerVM
```
Comment (vguzmanp) [inline Tests/Services/ViewModel/TestTeamsManagerVM.cs]: In this test you only assert, there isn't any call to anything in the system under test, is that intentional?

File: LongoMatch.Drawing/CanvasObjects/Teams/PlayersTaggerView.cs [hunk 14]
```diff
@@ -816,6 +542,75 @@ public void Move (Selection s, Point p, Point start)
...
+
+			if (ViewModel.AwayTeam != null) {
+				if ((ViewModel.NeedsSync (e.PropertyName, nameof (ViewModel.AwayTeam), sender, ViewModel) ||
+					 ViewModel.NeedsSync (e.PropertyName, nameof (ViewModel.AwayTeam.FieldPlayersList), sender, ViewModel.AwayTeam) ||
+				 ViewModel.NeedsSync (e.PropertyName, nameof (ViewModel.AwayTeam.BenchPlayersList), sender, ViewModel.AwayTeam))) {
+					LoadTeams ();
+					ReDraw ();
+				}
+				if (ViewModel.NeedsSync (e.PropertyName, nameof (ViewModel.AwayTeam.Tagged), sender, ViewModel.AwayTeam) ||
+					ViewModel.NeedsSync (e.PropertyName, nameof (ViewModel.AwayTeam.Tagged), sender, ViewModel)) {
```
Comment (vmartos) [inline LongoMatch.Drawing/CanvasObjects/Teams/PlayersTaggerView.cs]: This was only done to correctly synchronise the teams button state when setting the ViewModel in the view and call ViewModel.Sync(), because it produces a property changed event with the ViewModel as sender.

File: LongoMatch.Services/ViewModel/NewProjectPanelVM.cs [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+using LongoMatch.Core.Store;
+using LongoMatch.Core.ViewModel;
+
+namespace LongoMatch.Services.ViewModel
+{
+	public class NewProjectPanelVM : LMProjectVM
+	{
+		public NewProjectPanelVM ()
+		{
+			TeamTagger = new LMTeamTaggerVM ();
```
Comment (vmartos) [inline LongoMatch.Services/ViewModel/NewProjectPanelVM.cs]: http://stackoverflow.com/questions/3847832/understanding-private-setters/34223746#34223746

File: LongoMatch.Services/ViewModel/TeamsManagerVM.cs [hunk 16]
```diff
@@ -38,6 +38,17 @@ public TeamsManagerVM ()
 			DeleteCommand.Icon = Resources.LoadIcon ("longomatch-delete", StyleConf.TemplatesIconSize);
 			ExportCommand.Icon = Resources.LoadIcon ("longomatch-export", StyleConf.TemplatesIconSize);
 			ImportCommand.Icon = Resources.LoadIcon ("longomatch-import", StyleConf.TemplatesIconSize);
+			TeamTagger = new LMTeamTaggerVM ();
+			TeamTagger.Background = App.Current.HHalfFieldBackground;
```
Comment (vmartos) [inline LongoMatch.Services/ViewModel/TeamsManagerVM.cs]: I don't think this is a Typo, this was done on purpose because HalfFieldBackground already exists...

Comment (vmartos) [inline LongoMatch.Services/ViewModel/TeamsManagerVM.cs]: It was there before, I don't know why...

File: LongoMatch.Drawing/CanvasObjects/Teams/LMPlayerView.cs [hunk 17]
```diff
@@ -137,11 +121,12 @@ public override void Draw (IDrawingToolkit tk, Area area)
 			tk.DrawRectangle (p, size, 3);
 
 			/* Draw Arrow */
-			if (SubstitutionMode && (Highlighted || Active)) {
+			if (SubstitutionMode && (Highlighted || Player.Tagged)) {
 				ISurface arrow;
 				Point ap;
 
-				if (Playing) {
+				//FIXME: vmartos this is not working
```
Comment (vmartos) [inline LongoMatch.Drawing/CanvasObjects/Teams/LMPlayerView.cs]: I will fix it after processing all comments

File: LongoMatch.Drawing/Widgets/LMTeamTaggerView.cs [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+		}
+
+		public LMTeamTaggerView () : this (null)
+		{
+		}
+
+		protected override void DisposeManagedResources ()
+		{
+			tagger.Dispose ();
+			base.DisposeManagedResources ();
```
Comment (vmartos) [inline LongoMatch.Drawing/Widgets/LMTeamTaggerView.cs]: I don't think the order can influence the behaviour here

File: LongoMatch.GUI/Gui/Component/CodingWidget.cs [hunk 19]
```diff
@@ -279,59 +246,18 @@ void HandleCapturerTick (CapturerTickEvent e)
...
 				buttonswidget.CurrentTime = e.Time;
-				teamtagger.CurrentTime = e.Time;
 			}
 		}
 
 		void HandlePropertyChanged (object sender, PropertyChangedEventArgs e)
 		{
 			if (e.PropertyName == nameof (ViewModel.Capturer.CurrentCaptureTime)
-			    && ViewModel.Project.ProjectType == ProjectType.FileProject) {
+				&& ViewModel.Project.ProjectType == ProjectType.FileProject) {
```
Comment (vmartos) [inline LongoMatch.GUI/Gui/Component/CodingWidget.cs]: this was an automatic change, I think here that if the current Capture time is changed and is a FileProject do something

File: LongoMatch.GUI/Gui/Component/TeamTemplateEditor.cs [hunk 20]
```diff
@@ -391,5 +369,21 @@ void HandleColorSet (object sender, EventArgs e)
 			Edited = true;
 		}
 
+		void HandleTeamPropertyChanged (object sender, System.ComponentModel.PropertyChangedEventArgs e)
+		{
+			if (ViewModel.Team.NeedsSync (e.PropertyName, nameof (ViewModel.Team.Model),
+											  sender, ViewModel.Team)) {
+				Team = ViewModel.Team.Model as LMTeam;
+			}
+			if (ViewModel.Team.NeedsSync (e.PropertyName, $"Collection_{nameof (ViewModel.Team.Selection)}",
```
Comment (vmartos) [inline LongoMatch.GUI/Gui/Component/TeamTemplateEditor.cs]: No, At least modifying Selection by adding new players only Collection_Selection emits as property name. Only when we replace entirely the Selection property for another list it is emitted.

File: LongoMatch.GUI/Gui/Dialog/PlayEditor.cs [hunk 21]
```diff
@@ -38,10 +35,10 @@ namespace LongoMatch.Gui.Dialog
 {
 	// Fixme: Change the view to not use the model, use the VM provided
```
Comment (vmartos) [inline LongoMatch.GUI/Gui/Dialog/PlayEditor.cs]: This view has much more thinks other than a teamtagger. The entire view needs to be ported to MVVM I have only ported a portion of it.

File: LongoMatch.GUI/Gui/Dialog/SubstitutionsEditor.cs [hunk 22]
```diff
@@ -106,60 +104,19 @@ public void OnUnload ()
...
-			}
-			SwitchPlayer (substitution.In, substitution.Out);
-		}
-
 		protected override void OnResponse (ResponseType response_id)
 		{
 			base.OnResponse (response_id);
 			if (response_id == ResponseType.Ok) {
-				SaveChanges ();
+				ViewModel.SaveCommand.Execute ();
```
Comment (vmartos) [inline LongoMatch.GUI/Gui/Dialog/SubstitutionsEditor.cs]: This command can always be executed

File: LongoMatch.GUI/Gui/Panel/NewProjectPanel.cs [hunk 23]
```diff
@@ -400,6 +403,7 @@ void LoadTemplate (LMTeam template, TeamType team, bool forceColor)
 				} else {
 					homecolor1button.Click ();
 				}
+				ViewModel.TeamTagger.HomeTeam.Model = hometemplate;
```
Comment (vmartos) [inline LongoMatch.GUI/Gui/Panel/NewProjectPanel.cs]: Because this panel is not ported yet to MVVM

File: LongoMatch.GUI/Gui/Panel/TeamsTemplatesPanel.cs [hunk 24]
```diff
@@ -56,8 +56,8 @@ public TeamsTemplatesPanel ()
 
 			teamimage.Pixbuf = Helpers.Misc.LoadIcon ("longomatch-team-header", StyleConf.TemplatesHeaderIconSize);
 			playerheaderimage.Pixbuf = Helpers.Misc.LoadIcon ("longomatch-player-header", StyleConf.TemplatesHeaderIconSize);
-			newplayerimage.Pixbuf = Helpers.Misc.LoadIcon ("longomatch-add", StyleConf.TemplatesIconSize);
-			deleteplayerimage.Pixbuf = Helpers.Misc.LoadIcon ("longomatch-delete", StyleConf.TemplatesIconSize);
+			//newplayerimage.Pixbuf = Helpers.Misc.LoadIcon ("longomatch-add", StyleConf.TemplatesIconSize);
+			//deleteplayerimage.Pixbuf = Helpers.Misc.LoadIcon ("longomatch-delete", StyleConf.TemplatesIconSize);
```
Comment (vmartos) [inline LongoMatch.GUI/Gui/Panel/TeamsTemplatesPanel.cs]: because the buttons are now bind using the bindingContext with a command

File: LongoMatch.Services/Controller/LMTeamEditorController.cs [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+namespace LongoMatch.Services.Controller
+{
+	[Controller (TeamsManagerState.NAME)]
+	public class LMTeamEditorController : ControllerBase
+	{
+		LMTeamEditorVM teamEditor;
+
+		public override void SetViewModel (IViewModel viewModel)
+		{
+			teamEditor = (viewModel as ILMTeamEditorVM)?.TeamEditor;
```
Comment (vmartos) [inline LongoMatch.Services/Controller/LMTeamEditorController.cs]: changed to a cast

File: LongoMatch.Services/Controller/LMTeamTaggerController.cs [hunk 26]
```diff
@@ -0,0 +1,384 @@
...
+		}
+
+		public override void Start ()
+		{
+			base.Start ();
+			App.Current.EventsBroker.Subscribe<TagPlayerEvent> (HandleTagPlayerEvent);
+			App.Current.EventsBroker.Subscribe<UpdateLineup> (HandleUpdateLineup);
+			App.Current.EventsBroker.Subscribe<EventsDeletedEvent> (HandleEventsDeletedEvent);
+			App.Current.EventsBroker.Subscribe<EventEditedEvent> (HandleEventEdited);
+			UpdateLineup ();
```
Comment (vmartos) [inline LongoMatch.Services/Controller/LMTeamTaggerController.cs]: It should! If not it is a incorrect behaviour. The screen state first initialize then start controllers.

File: LongoMatch.Services/Controller/PlayEditorController.cs [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+				UpdatePlayTeams (teamTagger.AwayTeam);
+			}
+		}
+
+		void UpdatePlayTeams (LMTeamVM team)
+		{
+			if (team.Tagged && !playEditorVM.Play.Teams.Contains (team.Model)) {
+				playEditorVM.Play.Teams.Add (team.Model);
+			} else if (!team.Tagged && playEditorVM.Play.Teams.Contains (team.Model)) {
+				playEditorVM.Play.Teams.Remove (team.Model);
```
Comment (vmartos) [inline LongoMatch.Services/Controller/PlayEditorController.cs]: The playEditor is not fully migrated to MVVM

File: LongoMatch.Services/Controller/SubstitutionsEditorController.cs [hunk 28]
```diff
@@ -0,0 +1,110 @@
...
+	{
+		SubstitutionsEditorVM substitutionEditor;
+		LMPlayerVM taggedPlayer;
+
+		public override void Start ()
+		{
+			base.Start ();
+			App.Current.EventsBroker.Subscribe<UpdateEvent<SubstitutionEvent>> (HandleSaveSubstitutionEvent);
+			App.Current.EventsBroker.Subscribe<UpdateEvent<LineupEvent>> (HandleSaveLineupEvent);
+			if (substitutionEditor != null) {
```
Comment (vmartos) [inline LongoMatch.Services/Controller/SubstitutionsEditorController.cs]: This is something that we should standardise also, I think that the controllers should subscribe to VM property changes in start, and should unsubscribe on stop, the same way we do for events in eventbroker. because if an ViewModel is passed to another state we do not want to react to this property changes. when the controller is stopped

File: LongoMatch.Services/Controller/SubstitutionsEditorController.cs [hunk 29]
```diff
@@ -0,0 +1,110 @@
...
+
+		void SwitchPlayers ()
+		{
+			LMPlayerVM inOutPlayer = null;
+			if (substitutionEditor.InPlayer.Tagged) {
+				inOutPlayer = substitutionEditor.InPlayer;
+			}
+			if (substitutionEditor.OutPlayer.Tagged) {
+				inOutPlayer = substitutionEditor.OutPlayer;
+			}
```
Comment (vmartos) [inline LongoMatch.Services/Controller/SubstitutionsEditorController.cs]: no they can't

File: Tests/Services/ViewModel/TestTeamsManagerVM.cs (hunk 13 above)
Comment (vmartos) [inline Tests/Services/ViewModel/TestTeamsManagerVM.cs]: Yes this is only testing that Initialization is done correctly, because now it has some subVM's that needs to be correctly initialized to work as expected in views/controllers

File: Tests/Controller/TestLMTeamTaggerController.cs (hunk 12 above)
Comment (vmartos) [inline Tests/Controller/TestLMTeamTaggerController.cs]: If TeamTaggerController has a project it goes to a different code path, I testing that it behaves the same having a project or not.

File: LongoMatch.GUI/Gui/Panel/NewProjectPanel.cs [hunk 30]
```diff
@@ -654,28 +657,21 @@ void HandleShowMenuEvent (List<LMPlayer> players)
...
-					teamtagger.Reload ();
+					App.Current.EventsBroker.Publish (new UpdateLineup ());
 				};
 			} else {
 				item = new MenuItem ("Reset players");
 				item.Activated += (sender, e) => {
 					hometemplate.ResetPlayers ();
 					awaytemplate.ResetPlayers ();
-					teamtagger.Reload ();
+					App.Current.EventsBroker.Publish (new UpdateLineup ());
```
Comment (vmartos) [inline LongoMatch.GUI/Gui/Panel/NewProjectPanel.cs]: As replied before this view hasn't been ported to MVVM, maybe we should add techincal debt stories for all views that are not fully ported to MVVM, I can Add a fixme also

File: LongoMatch.Services/Controller/LMTeamTaggerController.cs [hunk 31]
```diff
@@ -0,0 +1,384 @@
...
+		KeyValuePair<PlayerVM, TeamVM> substitutionPlayer;
+		LMProjectVM project;
+		VideoPlayerVM videoPlayer;
+		bool isAnalysis;
+		Time lastTime;
+
+		List<LMPlayerVM> homeStartingPlayers = new List<LMPlayerVM> ();
+		List<LMPlayerVM> homeBenchPlayers = new List<LMPlayerVM> ();
+		List<LMPlayerVM> awayStartingPlayers = new List<LMPlayerVM> ();
+		List<LMPlayerVM> awayBenchPlayers = new List<LMPlayerVM> ();
```
Comment (vmartos) [inline LongoMatch.Services/Controller/LMTeamTaggerController.cs]: this is not the same as the properties in LMTeamVM, the properties in the viewmodel are the current field players and bench players based on the CurrentTime. This are just fields that are being calculated in SetLineup to get the starting field players list and bench players list based on the Lineup at the start of the match

File: LongoMatch.Services/Controller/LMTeamTaggerController.cs [hunk 32]
```diff
@@ -0,0 +1,384 @@
...
+		{
+			substitutionPlayer = new KeyValuePair<PlayerVM, TeamVM> ();
+		}
+
+		public override void SetViewModel (IViewModel viewModel)
+		{
+			TeamTagger = (viewModel as ILMTeamTaggerVM)?.TeamTagger;
+			var analysisVM = viewModel as IAnalysisViewModel;
+			if (analysisVM != null) {
+				isAnalysis = true;
```
Comment (vmartos) [inline LongoMatch.Services/Controller/LMTeamTaggerController.cs]: Actually if we pass a VM that has a VideoPlayer doesn't mean that is an analysis panel. this isAnalysis field is only used to know that when a substitution is made we should send a SubstitutionEvent. I can make a private property to make that, but then a cast should be performed every time I access the property, since we cannot change the Controllers ViewModels during execution only during initialization the ViewModel should change and i Think it can be calculated there.

File: LongoMatch.Services/ViewModel/NewProjectPanelVM.cs (hunk 15 above)
Comment (vguzmanp) [inline LongoMatch.Services/ViewModel/NewProjectPanelVM.cs]: nice

File: LongoMatch.Services/Controller/SubstitutionsEditorController.cs [hunk 33]
```diff
@@ -0,0 +1,110 @@
...
+				taggedPlayer = null;
+			}
+		}
+
+		void HandleViewModelPropertyChanged (object sender, System.ComponentModel.PropertyChangedEventArgs e)
+		{
+			if (!substitutionEditor.LineupMode) {
+				if (e.PropertyName == "Tagged") {
+					var player = sender as LMPlayerVM;
+					if (player != null) {
```
Comment (vmartos) [inline LongoMatch.Services/Controller/SubstitutionsEditorController.cs]: A cast should always performed anyway, I prefer this logic here

File: LongoMatch.Services/ViewModel/PlayEditorVM.cs (hunk 8 above)
Comment (vmartos) [inline LongoMatch.Services/ViewModel/PlayEditorVM.cs]: It don't crash since there exists a protection against null in FieldObject. And then it doesn't render any background.

File: LongoMatch.GUI/Gui/Dialog/PlayEditor.cs (hunk 21 above)
Comment (vguzmanp) [inline LongoMatch.GUI/Gui/Dialog/PlayEditor.cs]: 👌

File: LongoMatch.Services/ViewModel/PlayEditorVM.cs (hunk 7 above)
Comment (vmartos) [inline LongoMatch.Services/ViewModel/PlayEditorVM.cs]: This was used to cast in the LMTeamTaggerController, since uses the IVIdeoProjectVM to cast, i can make an "as" cast to ProjectVM first (to remove this), then if null use the interface to obtain the ProjectVM.

File: Tests/Controller/TestLMTeamTaggerController.cs [hunk 34]
```diff
@@ -94,11 +94,11 @@ public void TestTagPlayerSingleModeSubstitutionModeFalse ()
 			var clickedPlayer1 = teamTagger.HomeTeam.ViewModels.FirstOrDefault () as LMPlayerVM;
 			var clickedPlayer2 = teamTagger.HomeTeam.ViewModels [1] as LMPlayerVM;
 
-			teamTagger.PlayerClick (clickedPlayer1, false);
+			teamTagger.PlayerClick (clickedPlayer1, ButtonModifier.None);
 			Assert.IsTrue (clickedPlayer1.Tagged);
-			teamTagger.PlayerClick (clickedPlayer1, true);
+			teamTagger.PlayerClick (clickedPlayer1, ButtonModifier.Shift);
```
Comment (vguzmanp) [inline Tests/Controller/TestLMTeamTaggerController.cs]: What will happen with a different ButtonModifier?

File: LongoMatch.Drawing/CanvasObjects/Teams/LMPlayerView.cs [hunk 35]
```diff
@@ -58,6 +58,11 @@ public LMPlayerView ()
 			set;
 		}
 
+		public bool Playing {
```
Comment (vguzmanp) [inline LongoMatch.Drawing/CanvasObjects/Teams/LMPlayerView.cs]: I'm not entirely against this, but why is the one from the VM not working? I think it's better to have this in the VM, so that other views from the same VM get updated as well.

File: LongoMatch.GUI/Gui/Panel/NewProjectPanel.cs (hunk 30 above)
Comment (vguzmanp) [inline LongoMatch.GUI/Gui/Panel/NewProjectPanel.cs]: I think a TechDebt for that is too much overhead, but a FIXME is always welcome

File: LongoMatch.Services/Controller/PlayEditorController.cs (hunk 27 above)
Comment (vguzmanp) [inline LongoMatch.Services/Controller/PlayEditorController.cs]: Add a FIXME

File: LongoMatch.Services/Controller/SubstitutionsEditorController.cs (hunk 28 above)
Comment (vguzmanp) [inline LongoMatch.Services/Controller/SubstitutionsEditorController.cs]: Comment it in chat please

File: LongoMatch.Services/Controller/SubstitutionsEditorController.cs (hunk 29 above)
Comment (vguzmanp) [inline LongoMatch.Services/Controller/SubstitutionsEditorController.cs]: Then it should be an `else if`

File: LongoMatch.Drawing/CanvasObjects/Teams/LMPlayerView.cs (hunk 35 above)
Comment (vmartos) [inline LongoMatch.Drawing/CanvasObjects/Teams/LMPlayerView.cs]: At first i thought that Playing (in the Model/VM) was true if a player is in the field, but it was if the player is "Convocado" :). So accessing the Property Playing didn't work as expected. The Playing Property in The LMPLayerView is that the Player is in the field, so now is not a problem that the BenchObject View sets its player list to Playing to false and the FieldObject View sets its player list to Playing True.

File: LongoMatch.Services/State/NewProjectState.cs (hunk 4 above)
Comment (vmartos) [inline LongoMatch.Services/State/NewProjectState.cs]: this was called here:

 https://github.com/LongoMatch/longomatch/blob/LON-959/LongoMatch.Services/ToolsManager.cs#L128 

and was a passing a Model, so it was not working. The other call that was passing the LMProjectVM was here:

https://github.com/LongoMatch/longomatch/blob/LON-959/LongoMatch.Services/State/ProjectAnalysisState.cs#L59

That was passing a LMProjectVM, but since i changed the VM in the state to NewProjectVM, I decided to pass the Model and create the necessary ViewModel

Comment (vguzmanp) [inline LongoMatch.Services/State/NewProjectState.cs]: The problem with creating VM's is that if, for some reason, two states need to share the VM, they won't be able to. For example, in RiftAnalyst we share VM between the helper and the AnalysisView

File: LongoMatch.Services/Controller/LMTeamTaggerController.cs [hunk 36]
```diff
@@ -109,14 +109,14 @@ public LMTeamTaggerController ()
...
-			TeamTagger = (viewModel as ILMTeamTaggerVM)?.TeamTagger;
+			TeamTagger = ((ILMTeamTaggerDealer)viewModel).TeamTagger;
 			var analysisVM = viewModel as IAnalysisViewModel;
 			if (analysisVM != null) {
 				isAnalysis = true;
 				Project = (LMProjectVM)analysisVM.Project;
 				VideoPlayer = analysisVM.VideoPlayer;
 			} else {
-				Project = (viewModel as ILMProjectVM)?.Project;
+				Project = (LMProjectVM)(viewModel as IProjectDealer)?.Project;
```
Comment (vguzmanp) [inline LongoMatch.Services/Controller/LMTeamTaggerController.cs]: Are you sure this works?

if `viewModel` is a LMProjectVM the as cast will return null, and then the cast to LMProjectVM will fail

File: LongoMatch.Services/ViewModel/SubstitutionsEditorVM.cs [hunk 37]
```diff
@@ -94,18 +107,12 @@ public SubstitutionsEditorVM ()
 			protected set;
 		}
 
-		public LMProjectVM Project {
+		ProjectVM IProjectDealer.Project {
```
Comment (vguzmanp) [inline LongoMatch.Services/ViewModel/SubstitutionsEditorVM.cs]: So, if you need to do this, why not a ILMProjectDealer?

File: LongoMatch.Services/ViewModel/LMDrawingToolVM.cs [hunk 38]
```diff
@@ -0,0 +1,49 @@
...
+		}
+
+		void ResetTeamTagger ()
+		{
+			var project = Project as LMProject;
+			if (project != null) {
+				TeamTagger.AwayTeam.Model = project.VisitorTeamTemplate;
+				TeamTagger.HomeTeam.Model = project.LocalTeamTemplate;
+				TeamTagger.Background = project.Dashboard?.FieldBackground;
+			}
```
Comment (vguzmanp) [inline LongoMatch.Services/ViewModel/LMDrawingToolVM.cs]: Comment
